File: legacy/pasarguardbot/app/models/webapp/services.py
"""WebApp DTOs: services list/detail, config links, clients, link/subscription change."""

from pydantic import BaseModel, Field

from app.models.webapp.common import ServiceButtons, ServiceStatus


class WebAppChangeLinkRequest(BaseModel):
    """Request to change user's proxy link."""

    code: int = Field(..., description="Service code")
    init_data: str | None = Field(None, description="Telegram init data")
    session_token: str | None = Field(None, description="Session token from OTP login")


class WebAppChangeSubscriptionRequest(BaseModel):
    """Request to change user's subscription."""

    code: int = Field(..., description="Service code")
    init_data: str | None = Field(None, description="Telegram init data")
    session_token: str | None = Field(None, description="Session token from OTP login")


class WebAppServicesRequest(BaseModel):
    """Request for services list."""

    session_token: str | None = Field(None, description="Session token from login")
    init_data: str | None = Field(None, description="Telegram init data")
    page: int = Field(1, ge=1, description="Page number")
    limit: int = Field(10, ge=1, le=30, description="Items per page (max 30)")
    search: str | None = Field(None, description="Search by config name or service code")
    panel_code: int | None = Field(None, description="Filter services by panel (used with panel grouping)")


class PanelGroupItem(BaseModel):
    """A panel with the count of services the user owns in it."""

    panel_code: int
    panel_name: str
    service_count: int


class WebAppServicesResponse(BaseModel):
    """Response for services list endpoint."""

    ok: bool
    services: list[ServiceStatus] | None = None
    total: int = 0
    page: int = 1
    limit: int = 10
    total_pages: int = 0
    grouping_enabled: bool = False
    panel_groups: list[PanelGroupItem] | None = None
    error: str | None = None


class WebAppServiceDetailRequest(BaseModel):
    """Request for single service detail."""

    code: int = Field(..., description="Service code")
    session_token: str | None = Field(None, description="Session token from login")
    init_data: str | None = Field(None, description="Telegram init data")


class WebAppServiceDetailResponse(BaseModel):
    """Response for single service detail."""

    ok: bool
    service: ServiceStatus | None = None
    buttons: ServiceButtons | None = None
    error: str | None = None


class WebAppConfigLinkItem(BaseModel):
    index: int
    name: str
    url: str


class WebAppConfigLinksRequest(BaseModel):
    code: int = Field(..., description="Service code")
    page: int = Field(1, ge=1)
    limit: int = Field(500, ge=1, le=500)
    session_token: str | None = Field(None, description="Session token from login")
    init_data: str | None = Field(None, description="Telegram init data")


class WebAppConfigLinksResponse(BaseModel):
    ok: bool
    links: list[WebAppConfigLinkItem] = Field(default_factory=list)
    total: int = 0
    page: int = 1
    total_pages: int = 0
    error: str | None = None


class WebAppClientItem(BaseModel):
    created_at: int
    user_agent: str | None = None
    app_name: str | None = None
    version: str | None = None
    platform: str | None = None
    ip_address: str | None = None
    hwid: str | None = None


class WebAppClientsRequest(BaseModel):
    code: int = Field(..., description="Service code")
    session_token: str | None = Field(None, description="Session token from login")
    init_data: str | None = Field(None, description="Telegram init data")


class WebAppClientsResponse(BaseModel):
    ok: bool
    clients: list[WebAppClientItem] = Field(default_factory=list)
    error: str | None = None

File: legacy/pasarguardbot/frontend/src/types/webapp/services.ts
/** Mirrors app/models/webapp/services.py */
import type { WebAppAuthRequest, ServiceButtons, ServiceStatus } from "./common";

export interface WebAppChangeLinkRequest {
  code: number;
  init_data?: string | null;
  username?: string | null;
  password?: string | null;
  session_token?: string | null;
}

export interface WebAppChangeSubscriptionRequest {
  code: number;
  init_data?: string | null;
  username?: string | null;
  password?: string | null;
  session_token?: string | null;
}

export interface WebAppServicesRequest extends WebAppAuthRequest {
  page?: number;
  limit?: number;
  search?: string | null;
  panel_code?: number | null;
}

export interface PanelGroupItem {
  panel_code: number;
  panel_name: string;
  service_count: number;
}

export interface WebAppServicesResponse {
  ok: boolean;
  services?: ServiceStatus[] | null;
  total: number;
  page: number;
  limit: number;
  total_pages: number;
  grouping_enabled: boolean;
  panel_groups?: PanelGroupItem[] | null;
  error?: string | null;
}

export interface WebAppServiceDetailRequest extends WebAppAuthRequest {
  code: number;
}

export interface WebAppServiceDetailResponse {
  ok: boolean;
  service?: ServiceStatus | null;
  buttons?: ServiceButtons | null;
  error?: string | null;
}

export interface WebAppConfigLinkItem {
  index: number;
  name: string;
  url: string;
}

export interface WebAppConfigLinksRequest extends WebAppAuthRequest {
  code: number;
  page?: number;
  limit?: number;
}

export interface WebAppConfigLinksResponse {
  ok: boolean;
  links: WebAppConfigLinkItem[];
  total: number;
  page: number;
  total_pages: number;
  error?: string | null;
}

export interface WebAppClientItem {
  created_at: number;
  user_agent?: string | null;
  app_name?: string | null;
  version?: string | null;
  platform?: string | null;
  ip_address?: string | null;
  hwid?: string | null;
}

export interface WebAppClientsRequest extends WebAppAuthRequest {
  code: number;
}

export interface WebAppClientsResponse {
  ok: boolean;
  clients: WebAppClientItem[];
  error?: string | null;
}
